61-matamata-turtle/technical_specification.md

```diff
@@ -0,0 +1,300 @@
+which will provide the means to assess, process, and retry Kafka events that resulted
+in errors during their initial processing. This service will make use of the
+[Kafka DLQ](https://github.com/ghga-de/hexkit/pull/123) additions to `hexkit`.
+
+### Included/Required:
+- Implementation of the DLQ service (including its test suite)
+- `hexkit`: Add exception information to DLQ event headers
+- `hexkit`: Add preview functionality to `KafkaDLQSubscriber`
+- `hexkit`: Add preview dry-run functionality to `KafkaDLQSubscriber`
+
+### Optional:
+- Add event ID header in `hexkit`, consisting of the original topic, partition, and
+offset
+
+### Not included:
+- A web-based user interface
+
+## API Definitions:
+
+### RESTful/Synchronous:
+
+DLQ topics consist of the original topic name, the service abbreviation, and the fixed
+suffix `-dlq`. The latter could technically be omitted, but using it visually
+distinguishes dlq topics from others and making debugging easier.
+When interacting with DLQ topics via the API, however, the above structure does not need
+to be remembered: only the service abbreviation and the plain topic name.
+
+- `GET /services`
+  - *Returns a list of all configured services*
+  - Auth Header: internal token
+  - Response Body: array of strings (service names)
+  - Response Status: `200 OK`
+- `GET /topics/{service}`
+  - *Returns a list of all topics configured for the chosen service*
+  - Auth Header: internal token
+  - Response Body: array of strings (topic names)
+  - Response Status: `200 OK`
+- `GET /{service}/{topic}`
+  - *Returns the next events in the topic, not exceeding the configured maximum.*
+    *This is a preview of the events and does not impact event ordering within the DLQ.*
+    *Like the test endpoint below, this endpoint is idempotent.*
+  - Auth Header: internal token
+  - Response Body: array of event objects
+  - Response Status:
+    - `200 OK`: The preview was successful
+    - `401 Unauthorized`: auth error (not authenticated)
+- `GET /test/{service}/{topic}`
+  - *Returns the result of processing the next event in the specified topic. Nothing*
+    *is published and offsets are not affected. Calling the same test path multiple*
+    *times will return the same result.*
+  - Auth Header: internal token
+  - Response Body: the event that would be published, empty if event would be discarded.
+  - Response Status:
+    - `200 OK`: The event test was successful
+    - `422 Unprocessable Entity`: The non-empty request body is not a valid event. 
+    - `401 Unauthorized`: auth error (not authenticated)
+- `POST /{service}/{topic}`
+  - *Processes the next event in the topic, optionally publishing the supplied event*
+  - Auth Header: internal token
+  - Request Body: empty
+  - Response Body: empty or JSON representation of corrected event
+    - Need to consider whether checks should be added for fields like correlation ID.
```

Comment:
Shouldn't this go to the *request* body?

Reply:
Yes, that's a typo